### DISCUSSION TITLE
Secure context

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#32087] [Ideas]
It is paramount that Keycloak is hosted under a secure context to prevent man in the middle attacks from being able to obtain tokens, cookies, passwords, or any other secret sent between Keycloak and applications or user-agents.

[Secure Contexts](https://www.w3.org/TR/secure-contexts/) which is a proposed W3C specification defines what a secure context is, in summary only local requests or requests over an encrypted channel (`https`) is considered a secure context.

In addition some [user-agent APIs](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts) are only available in a secure context. Examples include the [Cookie Store API](https://developer.mozilla.org/en-US/docs/Web/API/Cookie_Store_API).

### Current behaviour of Keycloak

Keycloak supports both `http` and `https`, with an option to configure on a per-realm basis if `https` is required or not (for all communication, local network communication, or not at all).

When Keycloak is exposed through a reverse proxy there are a few different modes; encryption can be terminated at the proxy, where the proxy communicates with Keycloak over `http`. Requests can be re-encrypted by the proxy (using one certificate for the proxy, and another for Keycloak). Finally, encryption can be passed through by the proxy, where the proxy does not have access to the non-encrypted request. 

Keycloak currently sends cookies in two formats; one with the `SameSite` and secure flags set, and another (`_LEGACY`) cookies without the `SameSite` flag set. As an example if Keycloak is accessed over an internal IP address over `http` the user-agent will only send the `_LEGACY` cookies as this is not considered a secure context. An additional thing worth nothing is any client using OIDC or SAML from a different context and accessing Keycloak over a non secure context will not have access to cookies, as such the user session will not be available in this case.

All of this basically results in the ability to access Keycloak either over `https` or `http`, where the latter is only recommended in development and testing scenarios.

### Proposed new behaviour

Introduce a server-wide option to configure `https` requirement, where the options would be:

* All requests
* Require a secure context (`localhost` and `*.localhost` are permitted over `http`) (this would be the default)
* External network requests
* Not required

This should by done as a new `security profile` configuration option, to allow other similar security related policies to be applied to the server globally (examples could be requiring `https` for all external requests from Keycloak, including redirect-uris).

The ability to override the above on a per-realm basis would be deprecated and removed (Realm.isSslRequired).

At the moment Keycloak checks the `https` requirements all over the code-base, but as part of this change this would be handled by a filter applied to all requests. The filter would send a `html` error page if the accepted content-type is `html` or a `json error message` if the content type is `html`. Finally, for other content-types it would simply return a `400` error. An alternative could be to automatically redirect to `https` using [`HSTS`](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security).

For cookies the `_LEGACY` cookies would be removed, and replaced by a smarter setting of the `secure` flag using the server-wide `https` requirement, and would be set if the `https` requirements are configured to `all` or `secure context required`.

**Comment by stianst** [comment]
A proposal on how the configuration could look like:

```
kc.sh start --security-https=always|secure-context|external|disabled
```

There was also some thoughts on having some pre-defined security profiles, so you could do something like:

```
--security=strict|lax|none
```

**Reply by ahus1** [reply]
I'm not a fan of the current flag `Realm.isSslRequired` as it is duplicating something I should solve in my infrastructure: If I don't want Keycloak to communicate via TLS, I put a proxy in front of it and ensure firewall rules to prevent clients to connect via HTTP. That's all there is to it, and Keycloak should work with that.

If someone connects via HTTP, Keycloak should send out non-secure cookies. Keycloak has multiple server-side ways to determine if the clients connects via securely - there shouldn't be an additional need to configure things. 

1. If the `hostname` options is set, and it starts with `https`, assume TLS, and don't evaluate other options
2. If TLS is enabled, assume TLS, and don't evaluate other options
3. If proxy headers are configured -> Look at the protocol forwarded in the HTTP header, and determine TLS from that.

Not sure if I missed a case here. 

Once TLS is assumed, Keycloak will send out secure cookies, otherwise don't. 

* If TLS assumed, set a HTML attribute in the login page that TLS is assumed, so we can have JavaScript code to check if the page is running in a secure context. If it isn't, print a warning that there is a misconfiguration (Keycloak thinks it is served via TLS, while the JavaScript code knows it isn't a secure context). This could happend if you put a reverse proxy in front of Keycloak which accepts HTTP, and then forwards it via HTTPS to Keycloak. 
* If a WebAuthn appears in a non-secure context, show a warning immediately before the user acts.

**Reply by vmuzikar** [reply]
I have a similar perspective as @ahus1.

I won't be arguing now around the Security Profile as such, as that is a more generic concept and as such makes sense to me. What I see as a bit problematic is the HTTPS Security profile. IMHO it a partially clashes with something we already have: hostname config, and HTTP/HTTPS listeners config (assuming the Realm SSL config goes away). I reckon the HTTPS profile would just affect whether the HTTP or HTTPS listener is enabled by default (and consequently whether the TLS material needs to be provided). But how would the profile work with the hostname provider? What happens if the user does something like `--hostname=http://foobar.com --security-https=secure-context`? Does Keycloak fails to start with a validation error?

My main concern here is that the HTTPS profile might represent a bit too much unnecessary complexity (especially in cases of "mixed" HTTP and HTTPS like `--security-https=external`) and might be confusing to the user. The concepts of hostname and HTTP/HTTPS listeners are not really going away and this is an additional layer on top of it.

Hence I like @ahus1 counterproposal better as it seems quite simple to me – just limit the solution here to cookies only based on if we detected TLS is in use.

Keycloak already enforces HTTPS by default (the HTTP listener is disabled) and therefore is aligned with the secure-context by default (except of the local access but IMHO it's still better from the security perspective to completely disable HTTP than leave it enabled for selected requests only).

**Comment by stianst** [comment]
@jonkoops @ahus1 @vmuzikar FIY

**Comment by jonkoops** [comment]
> As an example if Keycloak is accessed over an internal IP address over http the user-agent will only send the _LEGACY cookies as this is not considered a secure context.

Note that local addresses such as `127.0.0.1` are also considered a secure context, however some are not e.g `0.0.0.0` (see [MDN docs](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure)). In regards to this proposal, I see several distinct topics here:

### Enforcement of a secure context

We need to start enforcing that users run Keycloak in a [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts). The primary reason for this is that web browsers increasingly gate-keep [certain APIs](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts) so that they are only available in a secure context. We need (and are already using some) of these APIs (clipboard, WebAuthN, crypto, etc), but we haven't documented this,  nor are not actively informing users this is not a supported way of running Keycloak.

My proposal would be to make a secure context required, without any ability for a user opt-out. This avoids a two-tier system where some features just don't work, it will simply be a hard 'no' to running an insecure configuration. We can help educate the users by documenting this change in the server configuration guide, and by adding notices to the various UIs when they have configured Keycloak in a way that would lead to an insecure context, for example:

![A screenshot showing an alert shown to the user in the administration console, notifying the user that they are running in an insecure context.](https://github.com/user-attachments/assets/199394fa-787b-43b8-9677-0147dfde1467)

This initiative can and should be handled entirely separate from all other topics, this is a change we can do now, without changing anything about the cookie policy or enforcement of HTTPS we have in place. It would be a change purely to educate the user of these new requirements, and allow us to streamline our code (e.g. use the Web Crypto API instead of rolling our own crypto for Keycloak JS).

### The removal of the `_LEGACY` cookies & enforcement of `Secure` attribute

The 'legacy' cookies were always a stop-gap measure to deal with lack of support for the `SameSite` attribute. I propose we remove this cookie variant entirely and make it a breaking change for the next major, without the possibility of restoring the previous behavior.

Regular cookies should always have the `Secure` attribute set to align with the enforcement of a secure context. Under no circumstances should cookies be set without having the `Secure` attribute. This will impact users only if they are running without HTTPS in production, which really should not be supported in the first place, or when using custom domains during local development that do not run on `localhost` or a `*.localhost` domain, which is easily mitigated.

### Improving TLS enforcement

I agree with Stian's assessment that we should not allow users to configure TLS on a per realm basis. This should be governed by a global setting for the entire deployment. If Keycloak is not behind a proxy and exposed publicly we should be doing everything to enforce a secure connection at all times.

For all of these changes, I think it makes more sense to bite the bullet and make this a breaking change as part of the next major.

**Reply by stianst** [reply]
I'm really not convinced we can and should block non-secure contexts. Folks often start out running Keycloak for dev or test purposes on Kubernetes, VMs, or whatever that ends up being exposed not on localhost, so would just make it harder for those folks to try things out or play around with Keycloak.

From the APIs you are referring to that pretty much only affects WebAuthN right now I'd say? The WebAuthN authenticator could for example easily check that a secure context is used (or required) and log an error if not.

Just because one or two smaller things require a secure context doesn't mean we should require it for everything, and I'm definitively talking about dev/test environments, not prod environments, which is where the idea of a security policy configuration comes in (secure by default, but allows open it up explicitly if needed).

**Comment by stianst** [comment]
Found another and much simpler way to solve the legacy cookie problems in https://github.com/keycloak/keycloak/pull/32130, so closing this discussion
